### PR TITLE
[WIP] Prevent iframe embedding by default

### PR DESCRIPTION
**Pull request by eternoendless**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | 1.7.3.x
| Description?  | Currently it's possible to embed a PrestaShop instance in an iframe, which exposes it to click-jacking attacks. This PR makes iframe embedding from another domain disabled by default.
| Type?         | bug fix
| Category?     | CO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | http://forge.prestashop.com/browse/BOOM-4917
| How to test?  | Put your shop's URL here: http://savanttools.com/test-frame/

Many thanks to [Arif Khan](http://www.andmp.com/) for having pointed this out!

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/8807)

### CODE REVIEW

**Comment by pehelwan** [comment]
Can you request a CVE or I shall do the same ?

**Comment by LittleBigDev** [comment]
Something is wrong with travis. Missing html elements.

**Comment by LittleBigDev** [comment]
:lgtm:

---

Reviewed 1 of 1 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/prestashop/prestashop/8807#-:-L6qGsB3niCH1s16F8ve:bnfp4nl)*

**Comment by LittleBigDev** [comment]
Rebase needed

**Comment by kpodemski** [comment]
hey @eternoendless 

what about all demos on Addons? :P

**Comment by eternoendless** [comment]
Oh thanks @kpodemski I hadn't thought of that 

🤔 

Maybe we should add an option to configure the allowed domains somewhere in the BO.

**Comment by jolelievre** [comment]
has this been forgotten? or is it purposely ignored?

**Comment by pehelwan** [comment]
I have blogged about it on Dzone go ahead and read it. I will push a commit
soon, thereby avoiding this scenario

**Review by payloadartist — APPROVED** [review]
The CSP and X-Frame-Options headers successfully block this attack from happening.

**Comment by jolelievre** [comment]
yes but from what I understand it is not that simple as the demos on Addons need to display PrestaShop in a frame, so we can't block it all the time

**Comment by pehelwan** [comment]
Can you elaborate?

**Comment by pehelwan** [comment]
Or you might as well want to implement a back end functionality which might
allow the Administrator user to change the frame or CSP permissions as per
feasibility if that's what you hinted at?

**Comment by jolelievre** [comment]
Yes, I think that's what @eternoendless suggested, it should be configurable it the BO
Maybe the default value would be to block any frame, but this could be overriden by a specified allowed domain
I lack knowledge about the addons platform actually, so I am not totally sure

**Comment by pehelwan** [comment]
Sure , the Administrator can simply change the nginx conf just for that ,
generally speaking things or frameworks as well as applications come out of
the box with this security feature enabled by default which should be done
in case of prestashop as well. That again emphasises on security by default
and enforced strictly.

File: classes/Tools.php [hunk 1]
```diff
@@ -2218,6 +2218,10 @@ public static function generateHtaccess($path = null, $rewrite_settings = null,
         fwrite($write_fd, "# .htaccess automaticaly generated by PrestaShop e-commerce open-source solution\n");
         fwrite($write_fd, "# http://www.prestashop.com - http://www.prestashop.com/forums\n\n");
 
+        // prevent click-jacking
+        fwrite($write_fd, "Header append Content-Security-Policy \"frame-ancestors 'self';\"\n");
+        fwrite($write_fd, "Header append X-Frame-Options SAMEORIGIN\n\n");
```
Comment (Quetzacoalt91) [inline classes/Tools.php:2223]: This should be surrounded by `<IfModule mod_headers.c>`, which already exist in this .htaccess.

Please move these lines further in the method, as all environment does not have the "header" module enabled (for instance, the docker images).

**Review by Quetzacoalt91 — CHANGES_REQUESTED** [review]

**Comment by molearnik** [comment]
I'd strongly recommend to think about adding "Security Enhancements" link to e.g. BO/Configure/Advanced Parameters/Security Enhancements where Administrator could enable HTTP response headers like:
```
X-Frame-Options: SAMEORIGIN
X-Content-Type-Options: nosniff 
X-XSS-Protection: "1; mode=block" 
Strict-Transport-Security "max-age=31536000; includeSubDomains" 
Cache-Control: no-cache='set-cookie'
```
Vulnerability scanners complain about omission of these headers by default. 
Sure, PrestaShop Administrators could do that by tweaking their web servers configurations.
However, 
1. That requires some technical knowledge and it is less convenient
2. .htaccess mechanism is not recommended for performance reasons (https://www.nginx.com/resources/wiki/start/topics/examples/likeapache-htaccess/#)
**Q: Would it be possible to add these headers via PrestaShop hooks like actionOutputHTMLBefore?**
3. Adding these optional headers from PrestaShop code would be cleaner and web server agnostic.
4. You could turn this into advantage over competition.
5. You could add additional security enhancement to the prestashop pages e.g. this script fragment:
https://www.owasp.org/index.php/Clickjacking_Defense_Cheat_Sheet#Best-for-now_Legacy_Browser_Frame_Breaking_Script

Regarding demonstrating AddOns: PrestaShop websites used for demonstration purposes could have some security options disabled (should the "X-Frame-Options: SAMEORIGIN" break the demo).

**Comment by marionf** [comment]
@eternoendless 

Does this PR is still relevant ?

**Comment by matks** [comment]
@PierreRambaud I think we should update this PR in the scope of our security reinforcement for PrestaShop

**Comment by PierreRambaud** [comment]
> @PierreRambaud I think we should update this PR in the scope of our security reinforcement for PrestaShop

Woot didn't see this notification.
I'm not sure it should be handle by PrestaShop installation, but why the host configuration. Otherwise we will not be able to use demo.prestashop.com anymore :roll_eyes:

**Comment by eternoendless** [comment]
Like [I said before](https://github.com/PrestaShop/PrestaShop/pull/8807#issuecomment-373039420):

> Maybe we should add an option to configure the allowed domains somewhere in the BO.

I'm closing this PR out of staleness, but @PierreRambaud as @matks said, this is something that could be done as a security feature.

**Pull request closed without merging**